### ISSUE TITLE
Add helper methods to App for registering new NotFound, Error & NotAllowed handler functions

### ISSUE THREAD

**Issue by geggleto** [slimphp/Slim#1337]
Taken from https://github.com/slimphp/Slim/issues/1336: "It might also be good to add some helper methods to the app notFound method, error method, and notAllowed method in \Slim\App would make this super-easy to read and understand; just for ease of use and readability."  [Lewiscowles1986] 

//cc @akrabat @silentworks @Lewiscowles1986 can you please tag discussion and v3 ty.

I am neutral on this... it is rather trival to already invoke these via the new PSR-7 methods... but could be useful for those who wish to migrate their code from v2 to v3.

**Comment by Lewiscowles1986** [comment]
Or to those like me that try to use the code from the docs for PSR-7 that doesn't work :+1: 

Also is it lost on anyone that this is 1337 :astonished:

**Comment by JoeBengalen** [comment]
Do you mean helper methods to define the error handlers? Or a helper method to call an error yourself from any part of you code.

**Comment by geggleto** [comment]
@JoeBengalen I think its the latter that makes the most sense.

**Comment by Lewiscowles1986** [comment]
Thanks guys!

## User Story

**As a** _PHP developer_
**I want to** _Be able to define custom behaviour for as many application scenario use-cases as possible, with as little effort as possible._
**So that** _I Spend less-time focusing on the technology, and am able to service as many request-scenarios as is possible for the applications I build._ 

> N.b I Think should be able to save time working on the lower-level `\Slim\Container`, or other implementation of system detail; and more time adding what I need. (This is a common user-story for me, I can be a tad focused on getting my end-result)

**Comment by geggleto** [comment]
@Lewiscowles1986 I can't come up with a valid use case for this... can you provide us a pseudo code example of what you want?

The only thing I could come up with, defeats the purpose of PSR-7. which would be a proxy to avoid the would $this['notAllowedHandler'] ...

**Comment by akrabat** [comment]
This is currently a 2-liner:

``` php
$app->get('/', function ($request, $response, $args) {

    $errorHandler = $this->notFoundHandler;
    return $errorHandler($request, $response);
});
```

Obviously, you can replace `$this->notFoundHandler` with `$this->notAllowedHandler` or `$this->errorHandler`.

We should document this. I'm not sure that it's worth writing additional helpers on App though.

**Comment by geggleto** [comment]
:+1: This should be documented... for that matter I don't know if the Container is explained well enough in the current documentation. This is a prime example of it though :)

**Comment by Lewiscowles1986** [comment]
> @Lewiscowles1986 I can't come up with a valid use case for this... can you provide us a pseudo code example of what you want?

Sure thing

[patch-1 by @LewisCowles1986](https://github.com/lewiscowles1986/Slim/blob/patch-1/Slim/App.php)

allows me to _(Working code using above patch)_

``` PHP
$app->notFound( function ($request, $response) {
    return $response->withStatus(404)
        ->withHeader('Content-Type', 'text/html')
        ->write('404: Resource "'.$request->getUri().'" not found');
});
```

which IMHO is more usable and consistent to the $app->{get|post|delete|...} syntax employed elsewhere.

I Can take it out of patch and move it to my own repo if you think Slim users won't be interested, or it isn't in keeping with your ideas, but essentially, it can be used for Slim n00bs like me, and even helps current and legacy slim v2 users to achieve very readable and purposeful, familiar code.

**Comment by akrabat** [comment]
Are you saying that it's the setting up of your own notFound handler that's too complicated rather than the raising of a notFound during dispatch?

**Comment by Lewiscowles1986** [comment]
Hi there! :smile: 

Unfortunately the code you added earlier, and the last comment are not at all what this issue is about.

**As a** PHP developer  
**I want to** Be able to define custom behaviour for as many application scenario use-cases as possible, with as little effort as possible. **_In this case add event handlers for `notFound`, `error`, and `notAllowed`.**_  
**So that** I Spend less-time focusing on the technology **_(Adding to container, telling it to use container, using different method to other handlers)**_, and am able to service as many request-scenarios as is possible for the applications I build **_(Okay so maybe over-dramatic, I could use `notFound`, `error` and `notAllowed`, but it's harder than it should be, the code looks a little bit of a mess to do something so simple, and it is inconsistent with other similar tasks when coding using Slim.)**_  

Currently to register the application to handle a get request

``` php
$app->get('/', function($request, $response) {
    $response->write('Yo there!');
    return $response;
});
```

To add custom 404 not-found-handler **_(current)**_

``` php
$c = $app->getContainer(); // lets ignore if we have app or are adding the a container pre-app, either way it works like this
$c['notFound'] = function ($c) {
    return function ($request, $response) use ($c) {
        return $c['response']
            ->withStatus(404)
            ->withHeader('Content-Type', 'text/html')
            ->write('404: Resource "'.$request->getUri().'" not found');
    };
};
```

What I am proposing _**(included in *__[this patch](https://github.com/lewiscowles1986/Slim/blob/patch-1/Slim/App.php)_**)***

``` php
$app->notFound( function ($request, $response) {
    return $response
        ->withStatus(404)
        ->withHeader('Content-Type', 'text/html')
        ->write('404: Resource "'.$request->getUri().'" not found');
});
```

So basically I am asking if adding helper function(s) would be acceptable to 
- reduce code
- make overriding built-in handlers simpler
- promote a consistent format when adding handlers to deal with application scenarios
- decrease unnecessary knowledge of implementation details such as containers unless necessary.

With additional benefits
- Less code to write
- Clearer intent
- similar method name, and addition to v2

> Apologies if this is still not clear, and if the use of the word error has cropped up anywhere, confusing matters.

**Comment by akrabat** [comment]
The user story bit didn't help at all. A clear answer to "Do you mean helper methods to define the error handlers? Or a helper method to call an error yourself from any part of you code." would have been more helpful.

I've updated the issue title.

**Comment by akrabat** [comment]
For reference, this is the diff that @Lewiscowles1986  is proposing: https://github.com/slimphp/Slim/compare/3.x...Lewiscowles1986:patch-1

My thoughts are that I'm generally fine with adding this functionality. It's not a major thing for me as I only ever set these up once in an application, so I don't mind the additional closure. It's not like adding a route which is a  very frequentl thing.

I'd like to see the methods called `registerErrorHandler`, `registerNotFoundHandler` and `registerNotAllowedHandler` to make it clear what they do.

Though, arguably, a single method called `registerHander($name, callable $handler)` would also work?

**Comment by kminek** [comment]
i'm against this change, i think App class should not be polluted with more helper methods. it should be clearly documented (with examples) how to overwrite those services in the container.

**Comment by silentworks** [comment]
As these are setters and Container Interop does not define setters this is a no for me.

If you were to decide to use Zend Service Manager or Aura.Di this would lead to broken methods in App.

**Comment by codeguy** [comment]
I'm with @silentworks. This needs to be done on the container.

**Comment by juriansluiman** [comment]
Currently the way the handlers are overwritten are pretty OK to me, as they are a service (which you are used to) which returns a callback which you are used to too. So it's a 1+1 in known components for me.

If the issue is about an easy way to handle 404 or 500 errors in your own app, [I've written about that before](https://juriansluiman.nl/article/156/app-notfound-for-slim-v3) and with a little bit of assumptions in your app, this is really easy to solve in userland.

**Comment by Lewiscowles1986** [comment]
juriansluiman agreed, and nice article, but I still think needlessly containerizing common use-case parts of code could be a mistake. It doesn't really matter, but it is much less readable; Anyway I seem to be the only person that thinks `->notFound(...` is more readable than `->getContainer()[...`
